swiftnav/src/coords/ellipsoid.rs

```diff
@@ -0,0 +1,221 @@
+            // less than one. By dividing by the larger of S or C we ensure that we do
+            // not divide by zero as only one of S or C should ever be zero.
+            //
+            // This incurs an extra division each iteration which the author was
+            // explicityl trying to avoid and it may be that this solution is just
+            // reverting back to the method of iterating on T directly, perhaps this
+            // bears more thought?
+
+            if s > c {
+                c /= s;
+                s = 1.0;
+            } else {
+                s /= c;
+                c = 1.0;
+            }
+
+            // Check for convergence and exit early if we have converged.
+            if (s - prev_s).abs() < 1e-16 && (c - prev_c).abs() < 1e-16 {
+                break;
+            }
+            prev_s = s;
+            prev_c = c;
+        }
+
+        a_n = (s * s + c * c).sqrt();
+        let latitude = 1.0_f64.copysign(ecef.z()) * (s / (e_c * c)).atan();
+        let height = (p * e_c * c + ecef.z().abs() * s - Self::A * e_c * a_n)
+            / (e_c * e_c * c * c + s * s).sqrt();
+        LLHRadians::new(latitude, longitude, height)
+    }
+
+    /// Rotate an [`ECEF`] vector into the a local [`NED`] vector
+    #[must_use]
+    fn ecef2ned(ecef: &ECEF, ref_ecef: &ECEF) -> NED {
+        let m = ecef2ned_matrix(Self::ecef2llh(ref_ecef));
+        NED::from_vector3(m * ecef.as_vector_ref())
+    }
+
+    /// Rotate a local [`NED`] vector into an [`ECEF`] vector
+    #[must_use]
+    fn ned2ecef(ned: &NED, ref_ecef: &ECEF) -> ECEF {
+        let m = ecef2ned_matrix(Self::ecef2llh(ref_ecef));
+        ECEF::from_vector3(m.transpose() * ned.as_vector_ref())
+    }
+
+    /// Calculate the local [`NED`] vector between two [`ECEF`] positions
+    #[must_use]
+    fn ecef2ned_d(ecef: &ECEF, ref_ecef: &ECEF) -> NED {
+        let temp_vector = ecef - ref_ecef;
+        Self::ecef2ned(&temp_vector, ref_ecef)
+    }
+
+    /// Calculate the [`AzimuthElevation`] from one [`ECEF`] position to another
+    ///
+    /// First the vector between the points is converted into the local North, East,
+    /// Down frame of the reference point. Then we can directly calculate the
+    /// azimuth and elevation.
+    #[must_use]
+    fn ecef2azel(ecef: &ECEF, ref_ecef: &ECEF) -> AzimuthElevation {
+        /* Calculate the vector from the reference point in the local North, East,
+         * Down frame of the reference point. */
+        let ned = Self::ecef2ned_d(ecef, ref_ecef);
+
+        let azimuth = ned.e().atan2(ned.n());
+        /* atan2 returns angle in range [-pi, pi], usually azimuth is defined in the
+         * range [0, 2pi]. */
+        let azimuth = if azimuth < 0.0 {
+            azimuth + 2.0 * std::f64::consts::PI
+        } else {
+            azimuth
+        };
+
+        let elevation = (-ned.d() / ned.as_vector_ref().norm()).asin();
+        AzimuthElevation::new(azimuth, elevation)
+    }
```

Comment:
Is there any reason these can't take in a `&self` arg rather than `&ECEF`?

Reply:
It would mean moving them into the separate types. At some point I'd like to allow specifying the ellipsoid to use for these conversion functions, so my first thought was to put the core transformation math into this Ellipsoid trait. I'm not sure how best to add the ability to specifying the ellipsoid, so I'm fine with moving these into the respective types.